cleanup-passes/typescript.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+'use strict';
+
+import * as fse from 'fs-extra';
+import * as path from 'path';
+import {execFile} from 'child_process';
+import {promisify} from 'util';
+
+import {register} from '../cleanup-pass';
+import {ElementRepo} from '../element-repo';
+import {makeCommit} from './util';
+
+const generatorPackageName = '@polymer/gen-typescript-declarations';
+const generatorSemver = '^0.3.6';
+const npmScriptName = 'update-types';
+const npmScriptCommand =
+    'rm -f *.d.ts **/*.d.ts && gen-typescript-declarations --outDir .';
+
+const execFilePromise = promisify(execFile);
+
+/**
+ * This pass updates an element repo to include typings generated by
+ * https://github.com/Polymer/gen-typescript-declarations/ and configures an
+ * "update-types" NPM script that can be run to re-generate them.
+ *
+ * Throws an error if the repo has a missing or invalid package.json or if npm
+ * or bower aren't globally installed.
+ */
+async function typescriptPass(element: ElementRepo): Promise<void> {
+  const packageJsonPath = path.join(element.dir, 'package.json');
+  let packageJson: {
+    scripts?: {[key: string]: string},
+    devDependencies?: {[key: string]: string},
+  };
+  try {
+    packageJson = await fse.readJson(packageJsonPath);
+  } catch {
+    throw new Error(
+        `Missing or invalid package.json for element: ${element.ghRepo.name}`);
+  }
+
+  if (packageJson.devDependencies === undefined) {
```

Comment:
You can be a little more declarative if you want with object spread:

```ts
try {
  packageJson = {
    scripts: {},
    devDependencies: {},
    ...await fse.readJson(packageJsonPath);
  };
} ...
```

Reply:
Nice, although since I'm not used to that pattern I think it would have taken me a sec to figure out what was going on.

I've added a new `NpmConfig` type (shared with the new package-json pass, see above) which declares these fields as undefined. TypeScript doesn't infer that they can't be undefined with the suggested snippet, so another reason I think I'll stick with mine.